lib/ansible/cli/adhoc.py

```diff
@@ -37,50 +26,58 @@ class AdHocCLI(CLI):
         this command allows you to define and run a single task 'playbook' against a set of hosts
     '''
 
-    def parse(self):
+    def init_parser(self):
         ''' create an options parser for bin/ansible '''
-
-        self.parser = CLI.base_parser(
-            usage='%prog <host-pattern> [options]',
-            runas_opts=True,
-            inventory_opts=True,
-            async_opts=True,
-            output_opts=True,
-            connect_opts=True,
-            check_opts=True,
-            runtask_opts=True,
-            vault_opts=True,
-            fork_opts=True,
-            module_opts=True,
-            basedir_opts=True,
-            desc="Define and run a single task 'playbook' against a set of hosts",
-            epilog="Some modules do not make sense in Ad-Hoc (include, meta, etc)",
-        )
+        self.parser = super(AdHocCLI, self).init_parser(usage='%prog <host-pattern> [options]',
+                                                        desc="Define and run a single task"
+                                                        " 'playbook' against a set of hosts",
+                                                        epilog="Some modules do not make sense in"
+                                                        " Ad-Hoc (include, meta, etc)")
+
+        self.parser = cli.runas_options(self.parser)
+        self.parser = cli.inventory_options(self.parser)
+        self.parser = cli.async_options(self.parser)
+        self.parser = cli.output_options(self.parser)
+        self.parser = cli.connect_options(self.parser)
+        self.parser = cli.check_options(self.parser)
+        self.parser = cli.runtask_options(self.parser)
+        self.parser = cli.vault_options(self.parser)
+        self.parser = cli.fork_options(self.parser)
+        self.parser = cli.module_options(self.parser)
+        self.parser = cli.basedir_options(self.parser)
```

Comment:
Haven't dug into the details of the PR yet, but at first glance, this section raised both eyebrows. First impression is that I have no idea what to expect this code to do. I can only assume that either: a) there is some unusual magic going on somewhere,  or b) this code is wrong.

Gut reaction is this is too much magic, but I haven't  dive into the requirement or the rest of the implementation yet. But figured I should mention it before I acclimate.

Reply:
Looking deeper, assuming this is accumulating options in self.parse, and modify_options is creating a new immutable-ish parser instance and returning the replacement?

Reply:
(Still first impressions stuff) guess output_options() etc are actually modifying self.parse by side effect and accumulating the options.

Related, could the options data be built up first instead of output_options()/connect_options() etc incrementally modifying the passed in self.parser?

(Thats not entirely related to the singleton stuff, but it could make it more clear when the parser instance is finalized and immutable-ish)


``` python
def vault_options(parser):
    parser.add_option('--vault-password-file', default=[], dest='vault_password_files',
                      help="vault password file", action="callback", callback=CLI.unfrack_paths, type='string')
    parser.add_option('--vault-id', default=[], dest='vault_ids', action='append', type='string',
                      help='the vault identity to use')

   self.parser = vault_options(self.parser)
```

to:

``` python

option_data = [{'options': ['--vault-password-file'], 'default': [], 'dest': 'vault_password_files', 
                           'help':"vault password file", 'action': 'callback', 'callback_id': 'UNFRACK_PATHS',
                           'type': 'string'},
                          {'options': ['--vault-id'], 'default': [], 'dest': 'vault_ids', 'action':'append', 'type': 'string',
                            'help': ''the vault identity to use'}]

...

parser = init_parser()    # or whatever constructs it
for option in option_data:
    # or likely, some helper method to setup the args and resolve callback_id to a callable
    parser.add_option(option)
```

ie, pass all the data to the parser constructor instead of incrementally modifying it.

If there is nothing dynamic in the options data, it could be copied around and each cli could build it's
own parser instance.

Could potentially make it easier to build cli docs as well.

Reply:
Yeah, I kept this in a separate commit because I'm not sure I like it.  however, I think my disgruntlement is a little different.

The original code was bad because it was a poor organization of code.  A large function which has a huge number of parameters and a multiplication of potential output based on those options and there's conditionals gating behaviour for each one.

This code fixes that by splitting it up into separate functions.  The conditionals are no longer even present in the code because those have all been pre-determined when the CLI script is written so there's no need for any run-time decision making.  That's the good side of this...

The bad side is that we're modifying data by side effect.  Pass a parser in and the parser has new data added onto it over and over until we've built up the thing we want.  I could go whole hog on that and eliminate all the return values but of course... I don't like operating by side effect either.

I don't like defining a new format (dict with special fields) to hold the data either.  Feels a little fragile (I think that's a false feeling, though... it shouldn't be any more fragile than specifying the args directly) and wasteful (since there's already a format there).  I've looked at the code and documentation for OptParse and Argparse and it looks like optparse does let you create standalone Options() and then pass it into add_option().  It's not documented, though.  argparse, I had to look at the code but then I came to understand that argparse.Action is the equivalent in this case and that is explicitly documented.  So that gets us close without defining our own format.

The only remaining piece I see is that we add both options and optiongroups so I think we'll have to use isinstance to separate out options and optiongroups from each other when we actually add them.

So the new calling code should something like this:
``` python
options = []  # Can contain optiongroups as well
options.extend(vault_options())
options.extend(module_options())
self.parser = create_parser(usage, description, epilog, options)  # create_parser will divide options from optiongroups appropriately
```
Options custom to each script could now be done by adding the option to the ```options``` list before calling create_parser or they could continue to use self.parser.addOption() after the parser is created.

Reply:
So it seems like we have two choices:

(A) Go all in on a side-effect driven API:
``` python
parser = create_base_parser(usage, description, epilog)
add_vault_options(parser)
add_vault_rekey_options(parser)
self.parser = parser

def create_base_parser([..]):
    parser = OptionParser([...])
    return parser
```

(B) Create at least a partial format of our own to build up the data before creating it:
``` python
options = []
options = vault_options()
options.extend(vault_rekey_options)
self.parser = create_parser(options)

def create_parser([..]):
    parser = OptionParser([...])
    for option in options:
        if isinstance(option, MutableMapping):
            parser.add_option_group(option['desc'], option[..])
            for real_option in options:
                parser.add_option(real_option)
        else:
            parser.add_option(real_option)
    return parser
```

As much as I don't like side-effects, I think that option (A) is the winner here because the API and code we have to write is much simpler.  Getting rid of the functional-programming aspect of returning parser and renaming the functions should hopefully clarify how the API is to be used.  option (B) just ends up with us fighting the design of the Optparse API.  So even though it would be nicer if Optparse gave us a way to program it in a more functional style, using side effect seems to be the only natural way to program it.

I'll go ahead and make those modifications now.

Reply:
@alikins, okay, I've posted an update that makes it more clear that this code operates by side-effect.  the unittests won't pass yet because they are too close to the internals.  I'll work on those before combining everything back into a nice, neat set of commits.